File: backends/node/shared/implementations/middleware/request-enrichment/default/src/request-enrichment-options.ts
/** Proxy headers that the IP resolver can trust. */
export type TrustedProxyHeader = "cf-connecting-ip" | "x-real-ip" | "x-forwarded-for";

/**
 * Configuration options for request enrichment.
 * Mirrors D2.Shared.RequestEnrichment.Default.RequestEnrichmentOptions in .NET.
 */
export interface RequestEnrichmentOptions {
  /** Whether to perform WhoIs lookups for non-localhost IPs. Default: true. */
  enableWhoIsLookup: boolean;
  /** Header name for client-provided fingerprint. Default: "x-client-fingerprint". */
  clientFingerprintHeader: string;
  /**
   * Which proxy headers to trust for IP resolution.
   * Only headers in this list will be checked — others are ignored.
   * Default: ["cf-connecting-ip"] (Cloudflare only).
   *
   * Set to all three to trust any proxy:
   * ["cf-connecting-ip", "x-real-ip", "x-forwarded-for"]
   */
  trustedProxyHeaders: readonly TrustedProxyHeader[];
  /** Max length for client fingerprint header values. Values exceeding this are truncated. Default: 256. */
  maxFingerprintLength: number;
  /** Max length for forwarded-for / IP headers. Values exceeding this are truncated. Default: 2048. */
  maxForwardedForLength: number;
}

/** Default request enrichment options. */
export const DEFAULT_REQUEST_ENRICHMENT_OPTIONS: RequestEnrichmentOptions = {
  enableWhoIsLookup: true,
  clientFingerprintHeader: "x-client-fingerprint",
  trustedProxyHeaders: ["cf-connecting-ip"],
  maxFingerprintLength: 256,
  maxForwardedForLength: 2048,
};

File: backends/dotnet/shared/Implementations/Middleware/RequestEnrichment.Default/RequestEnrichmentOptions.cs
// -----------------------------------------------------------------------
// <copyright file="RequestEnrichmentOptions.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.RequestEnrichment.Default;

/// <summary>
/// Proxy headers that the IP resolver can trust.
/// </summary>
public enum TrustedProxyHeader
{
    /// <summary>Cloudflare CF-Connecting-IP header.</summary>
    CfConnectingIp,

    /// <summary>Nginx/reverse proxy X-Real-IP header.</summary>
    XRealIp,

    /// <summary>Standard X-Forwarded-For header.</summary>
    XForwardedFor,
}

/// <summary>
/// Configuration options for request enrichment middleware.
/// </summary>
public class RequestEnrichmentOptions
{
    /// <summary>
    /// Gets or sets a value indicating whether to perform WhoIs lookups.
    /// </summary>
    /// <remarks>
    /// When enabled, the middleware will call the Geo service to resolve
    /// geographic information for the client IP address.
    /// Default is true.
    /// </remarks>
    public bool EnableWhoIsLookup { get; set; } = true;

    /// <summary>
    /// Gets or sets the header name for client-provided fingerprint.
    /// </summary>
    /// <remarks>
    /// Default is "X-Client-Fingerprint".
    /// </remarks>
    public string ClientFingerprintHeader { get; set; } = "X-Client-Fingerprint";

    /// <summary>
    /// Gets or sets the proxy headers trusted for IP resolution.
    /// </summary>
    /// <remarks>
    /// Only headers in this set are checked — others are ignored.
    /// Default: Cloudflare only (CfConnectingIp).
    /// Set to all three to trust any proxy.
    /// </remarks>
    public HashSet<TrustedProxyHeader> TrustedProxyHeaders { get; set; } =
        [TrustedProxyHeader.CfConnectingIp];
}
